src/rules/no-unused-modules.js

```diff
@@ -0,0 +1,706 @@
+          localImport = new Set([currentValue])
+        }
+        imports.set(reexport.path, localImport)
+      })
+
+      localImportList.forEach((value, key) => {
+        if (isNodeModule(key)) {
+          return
+        }
+        imports.set(key, value.importedSpecifiers)
+      })
+      importList.set(file, imports)
+      
+      // build up export list only, if file is not ignored
+      if (ignoredFiles.has(file)) {
+        return
+      }
+      namespace.forEach((value, key) => {
+        if (key === DEFAULT) {
+          exports.set(IMPORT_DEFAULT_SPECIFIER, { whereUsed: new Set() })
+        } else {
+          exports.set(key, { whereUsed: new Set() })
+        }
+      })
+    }
+    exports.set(EXPORT_ALL_DECLARATION, { whereUsed: new Set() })
+    exports.set(IMPORT_NAMESPACE_SPECIFIER, { whereUsed: new Set() })
+    exportList.set(file, exports)
+  })
+  exportAll.forEach((value, key) => {
+    value.forEach(val => {
+      const currentExports = exportList.get(val)
+      const currentExport = currentExports.get(EXPORT_ALL_DECLARATION)
+      currentExport.whereUsed.add(key)
+    })
+  })
+}
+
+/**
+ * traverse through all imports and add the respective path to the whereUsed-list 
+ * of the corresponding export
+ */
+const determineUsage = () => {
+  importList.forEach((listValue, listKey) => {
+    listValue.forEach((value, key) => {
+      const exports = exportList.get(key)
+      if (typeof exports !== 'undefined') {
+        value.forEach(currentImport => {
+          let specifier
+          if (currentImport === IMPORT_NAMESPACE_SPECIFIER) {
+            specifier = IMPORT_NAMESPACE_SPECIFIER
+          } else if (currentImport === IMPORT_DEFAULT_SPECIFIER) {
+            specifier = IMPORT_DEFAULT_SPECIFIER
+          } else {
+            specifier = currentImport
+          }
+          if (typeof specifier !== 'undefined') {
+            const exportStatement = exports.get(specifier)
+            if (typeof exportStatement !== 'undefined') {
+              const { whereUsed } = exportStatement
+              whereUsed.add(listKey)
+              exports.set(specifier, { whereUsed })
+            }
+          }
+        })
+      }
+    })
+  })
+}
+
+const getSrc = src => {
+  if (src) {
+    return src
+  }
+  return [process.cwd()]
+}
+
+/**
+ * prepare the lists of existing imports and exports - should only be executed once at
+ * the start of a new eslint run
+ */
+const doPreparation = (src, ignoreExports, context) => {
+  const srcFiles = resolveFiles(getSrc(src), ignoreExports)
+  prepareImportsAndExports(srcFiles, context)
+  determineUsage()
+  preparationDone = true
+}
+
+const newNamespaceImportExists = specifiers =>
+  specifiers.some(({ type }) => type === IMPORT_NAMESPACE_SPECIFIER)
+
+const newDefaultImportExists = specifiers =>
+  specifiers.some(({ type }) => type === IMPORT_DEFAULT_SPECIFIER)
+
+module.exports = {
+  meta: {
+    docs: { url: docsUrl('no-unused-modules') },
```

Comment:
Isn't the name `no-unused-modules` misleading? This rule reports individual export declarations, but the name suggests it reports whole files as unused.

Reply:
my understanding is that it reports on both - both unused files, and unused exports.

Reply:
Just as @ljharb says, it reports on both:

- modules without any exports are reported [here](https://github.com/rfermann/eslint-plugin-import/blob/d35b7ff37c2778ed43d824cccf9553cd9faafc79/src/rules/no-unused-modules.js#L283)
- and modules with unused exports are reported [here](https://github.com/rfermann/eslint-plugin-import/blob/d35b7ff37c2778ed43d824cccf9553cd9faafc79/src/rules/no-unused-modules.js#L325) and [here](https://github.com/rfermann/eslint-plugin-import/blob/d35b7ff37c2778ed43d824cccf9553cd9faafc79/src/rules/no-unused-modules.js#L331)

However this rule currently does not report, if a module only consists of unused exports. Each of these exports is reported separately. 
I am not sure whether it makes sense to have one single report being triggered if none of the existing exports is used.

Reply:
It does make sense to me in that case to only provide one report; just as I’d expect when a module has no exports.